### PR TITLE
Remove 'strict mode' from parsing and checking, and make it purely something purely checked at bind time.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -531,6 +537,48 @@ namespace ts {
             typeLiteralSymbol.members = { [symbol.name]: symbol };
         }
 
+        function bindObjectLiteralExpression(node: ObjectLiteralExpression) {
+            if (inStrictMode) {
+                let seen: Map<number> = {};
+                const Property = 1;
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: It would be nice to comment that you picked 1 & 2 for the sake of truthiness.

Also, consider using a local const enum.

Comment (JsonFreeman) [inline src/compiler/binder.ts]: +1 for the const enum

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -88,6 +88,12 @@ namespace ts {
         let container: Node;
         let blockScopeContainer: Node;
         let lastContainer: Node;
+
+        // If this file is an external module, then it is automatically in strict-mode according to
+        // ES6.  If it is not an external module, then we'll determine if it is in strict mode or 
+        // not depending on if we see "use strict" in certain places.
```
Comment (JsonFreeman) [inline src/compiler/binder.ts]: Or if we see a class

File: src/compiler/binder.ts [hunk 3]
```diff
@@ -531,6 +537,48 @@ namespace ts {
             typeLiteralSymbol.members = { [symbol.name]: symbol };
         }
 
+        function bindObjectLiteralExpression(node: ObjectLiteralExpression) {
+            if (inStrictMode) {
+                let seen: Map<number> = {};
+                const Property = 1;
+                const NonProperty = 2;
```
Comment (JsonFreeman) [inline src/compiler/binder.ts]: MethodOrAccessor

File: src/compiler/binder.ts [hunk 4]
```diff
@@ -563,20 +611,135 @@ namespace ts {
...
+                    // We check first if the name is inside class declaration or class expression; if so give explicit message
+                    // otherwise report generic error message.
+                    let span = getErrorSpanForNode(file, name);
+                    file.bindDiagnostics.push(createFileDiagnostic(file, span.start, span.length,
+                        getStrictModeEvalOrArgumentsMessage(contextNode), identifier.text));
                 }
             }
         }
 
+        function getStrictModeEvalOrArgumentsMessage(node: Node) {
```
Comment (JsonFreeman) [inline src/compiler/binder.ts]: Share logic with getStrictModeIdentifierMessage?

File: src/compiler/binder.ts [hunk 5]
```diff
@@ -601,19 +769,78 @@ namespace ts {
...
+
+                if (isUseStrictPrologueDirective(statement)) {
+                    inStrictMode = true;
+                    return;
+                }
+            }
         }
         
+        /// Should be called only on prologue directives (isPrologueDirective(node) should be true)
+        function isUseStrictPrologueDirective(node: Node): boolean {
```
Comment (JsonFreeman) [inline src/compiler/binder.ts]: Can this take ExpressionStatement instead of node?

File: src/compiler/binder.ts [hunk 6]
```diff
@@ -601,19 +769,78 @@ namespace ts {
...
+
                 case SyntaxKind.TypeParameter:
                     return declareSymbolAndAddToSymbolTable(<Declaration>node, SymbolFlags.TypeParameter, SymbolFlags.TypeParameterExcludes);
                 case SyntaxKind.Parameter:
                     return bindParameter(<ParameterDeclaration>node);
-                case SyntaxKind.VariableDeclaration:
                 case SyntaxKind.BindingElement:
-                    return bindVariableDeclarationOrBindingElement(<VariableDeclaration | BindingElement>node);
+                case SyntaxKind.VariableDeclaration:
+                    return bindVariableDeclarationOrBindingElement(<BindingElement | VariableDeclaration>node);
```
Comment (JsonFreeman) [inline src/compiler/binder.ts]: I'd revert this change

File: src/compiler/parser.ts [hunk 7]
```diff
@@ -3913,16 +3879,15 @@ namespace ts {
...
+            return isIdentifier() || token === SyntaxKind.OpenBraceToken || token === SyntaxKind.OpenBracketToken;
         }
 
         function isLetDeclaration() {
-            // It is let declaration if in strict mode or next token is identifier\open bracket\open curly on same line.
-            // otherwise it needs to be treated like identifier
-            return inStrictModeContext() || lookAhead(nextTokenIsIdentifierOrStartOfDestructuringOnTheSameLine);
+            // In ES6 'let' always starts a lexical declaration if followed by an identifier or { 
+            // or [.
+            return lookAhead(nextTokenIsIdentifierOrStartOfDestructuring);
```
Comment (JsonFreeman) [inline src/compiler/parser.ts]: Add a test for this if you haven't already.

File: src/compiler/parser.ts [hunk 8]
```diff
@@ -2665,7 +2626,7 @@ namespace ts {
                 // for now we just check if the next token is an identifier.  More heuristics
                 // can be added here later as necessary.  We just need to make sure that we
                 // don't accidently consume something legal.
-                return lookAhead(nextTokenIsIdentifierOnSameLine);
+                return lookAhead(nextTokenIsIdentifierOrKeywordOrNumberOnSameLine);
```
Comment (JsonFreeman) [inline src/compiler/parser.ts]: Update comment above.

Comment (JsonFreeman) [inline src/compiler/parser.ts]: What about
```ts
for (yield of foo) { }
```
Do we do the right thing here? Same question for `yield instanceOf foo` and `yield in foo`. Does yield have low enough precedence to not even get here?

**Comment by JsonFreeman** [comment]
:+1:

File: src/compiler/binder.ts (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/compiler/binder.ts]: Will fix!

File: src/compiler/binder.ts (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/compiler/binder.ts]: Will do!

File: src/compiler/binder.ts (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/compiler/binder.ts]: I couldn't figure out a way.  Suggestions? :)

File: src/compiler/binder.ts (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/compiler/binder.ts]: Sure!

Comment (CyrusNajmabadi) [inline src/compiler/binder.ts]: Sure!

File: src/compiler/binder.ts (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/compiler/binder.ts]: Will do!

File: src/compiler/parser.ts (hunk 8 above)
Comment (CyrusNajmabadi) [inline src/compiler/parser.ts]: Those are good points.  Will investigate.  Thanks!

File: src/compiler/binder.ts (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/compiler/binder.ts]: Sure!

File: src/compiler/binder.ts (hunk 4 above)
Comment (JsonFreeman) [inline src/compiler/binder.ts]: I started to write it out, and then realized that while it is possible, it seems really silly. It's fine how it is.
